azmq fails to compile with boost 1.66 due to changes in asio

Hi,

Just tried upgrading my project which uses azmq to user boost 1.66 and there are some breaking changes in the asio library:

```cpp
In file included from /Users/foo/distrib/include/azmq/context.hpp:12:
/Users/foo/distrib/include/azmq/detail/socket_service.hpp:328:35: error: 'azmq::detail::socket_service::shutdown' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
        boost::system::error_code shutdown(implementation_type & impl,
                                  ^
/Users/foo/distrib/include/boost/asio/io_context.hpp:800:32: note: hidden overloaded virtual function 'boost::asio::io_context::service::shutdown' declared here: different number
      of parameters (0 vs 3)
  BOOST_ASIO_DECL virtual void shutdown();
                               ^
/Users/foo/distrib/include/azmq/socket.hpp:109:35: error: use of undeclared identifier 'implementation'
        if (get_service().do_open(implementation, type, optimize_single_threaded, ec))
                                  ^
/Users/foo/distrib/include/azmq/socket.hpp:115:38: error: use of undeclared identifier 'implementation'
        get_service().move_construct(implementation,
                                     ^
/Users/foo/distrib/include/azmq/socket.hpp:117:44: error: no member named 'implementation' in 'azmq::socket'
                                     other.implementation);
                                     ~~~~~ ^
/Users/foo/distrib/include/azmq/socket.hpp:121:35: error: use of undeclared identifier 'implementation'
        get_service().move_assign(implementation,
                                  ^
/Users/foo/distrib/include/azmq/socket.hpp:123:39: error: no member named 'implementation' in 'azmq::socket'
                                  rhs.implementation);
                                  ~~~ ^
```

Also, on a side note, it would be great if you could more regularly version (using a tag) the azmq library so a specific code level can more easily be recoverable from github.

Thanks for the library and all the effort you've put into it.

O.

There is one (important) bug fix since I tagged 1.0, it’s a fair point and
I will tag that in the next few days.

I’m aware there are breaking changes in 1.66 as Chris Kolhoff has updated
Asio to align more closely with the Networking TS, but I haven’t yet had
time to really look into the changes.

 I will see if I can get some time to address this in the next couple of
weeks. Patches are also welcome, if you woul like to have a go at it :)

Thanks,
Tom.

On Thu, Dec 28, 2017 at 5:26 AM ogrant <notifications@github.com> wrote:

> Hi,
>
> Just tried upgrading my project which uses azmq to user boost 1.66 and
> there are some breaking changes in the asio library:
>
> In file included from /Users/foo/distrib/include/azmq/context.hpp:12:
> /Users/foo/distrib/include/azmq/detail/socket_service.hpp:328:35: error: 'azmq::detail::socket_service::shutdown' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
>         boost::system::error_code shutdown(implementation_type & impl,
>                                   ^
> /Users/foo/distrib/include/boost/asio/io_context.hpp:800:32: note: hidden overloaded virtual function 'boost::asio::io_context::service::shutdown' declared here: different number
>       of parameters (0 vs 3)
>   BOOST_ASIO_DECL virtual void shutdown();
>                                ^
> /Users/foo/distrib/include/azmq/socket.hpp:109:35: error: use of undeclared identifier 'implementation'
>         if (get_service().do_open(implementation, type, optimize_single_threaded, ec))
>                                   ^
> /Users/foo/distrib/include/azmq/socket.hpp:115:38: error: use of undeclared identifier 'implementation'
>         get_service().move_construct(implementation,
>                                      ^
> /Users/foo/distrib/include/azmq/socket.hpp:117:44: error: no member named 'implementation' in 'azmq::socket'
>                                      other.implementation);
>                                      ~~~~~ ^
> /Users/foo/distrib/include/azmq/socket.hpp:121:35: error: use of undeclared identifier 'implementation'
>         get_service().move_assign(implementation,
>                                   ^
> /Users/foo/distrib/include/azmq/socket.hpp:123:39: error: no member named 'implementation' in 'azmq::socket'
>                                   rhs.implementation);
>                                   ~~~ ^
>
> Also, on a side note, it would be great if you could more regularly
> version (using a tag) the azmq library so a specific code level can more
> easily be recoverable from github.
>
> Thanks for the library and all the effort you've put into it.
>
> O.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zeromq/azmq/issues/133>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYBz2dmvGUS38pKW3WlISyN9G2_sUcks5tE3ragaJpZM4ROPas>
> .
>

No problem, thanks.

On the other hand, I am still seeing a compilation error:

In file included from include/azmq/socket.hpp:14:
In file included from include/azmq/context.hpp:12:
include/azmq/detail/socket_service.hpp:329:35: error: 'azmq::detail::socket_service::shutdown' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
        boost::system::error_code shutdown(implementation_type & impl,
                                  ^   
include/boost/asio/io_context.hpp:800:32: note: hidden overloaded virtual function 'boost::asio::io_context::service::shutdown' declared here: different number
      of parameters (0 vs 3)
  BOOST_ASIO_DECL virtual void shutdown();
                               ^   
1 error generated.

Regards,
O.

let me look at that.

On Thu, Jan 25, 2018 at 2:22 AM, ogrant <notifications@github.com> wrote:

> No problem, thanks.
>
> On the other hand, I am still seeing a compilation error:
>
> In file included from include/azmq/socket.hpp:14:
> In file included from include/azmq/context.hpp:12:
> include/azmq/detail/socket_service.hpp:329:35: error:
> 'azmq::detail::socket_service::shutdown' hides overloaded virtual
> function [-Werror,-Woverloaded-virtual]
> boost::system::error_code shutdown(implementation_type & impl,
> ^
> include/boost/asio/io_context.hpp:800:32: note: hidden overloaded virtual
> function 'boost::asio::io_context::service::shutdown' declared here:
> different number
> of parameters (0 vs 3)
> BOOST_ASIO_DECL virtual void shutdown();
> ^
> 1 error generated.
>
> Regards,
> O.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/zeromq/azmq/issues/133#issuecomment-360394170>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAHYByfNIHp1YAvv_HpAuFVV2dxJezCvks5tODmqgaJpZM4ROPas>
> .
>

--- a/azmq/detail/socket_service.hpp
+++ b/azmq/detail/socket_service.hpp
@@ -326,9 +326,9 @@ namespace detail {
             return ec;
         }
 
-        boost::system::error_code shutdown(implementation_type & impl,
-                                           shutdown_type what,
-                                           boost::system::error_code & ec) {
+        boost::system::error_code descriptor_shutdown(implementation_type & impl,
+                                                      shutdown_type what,
+                                                      boost::system::error_code & ec) {
             unique_lock l{ *impl };
             if (impl->shutdown_ < what)
                 impl->shutdown_ = what;

--- a/azmq/socket.hpp
+++ b/azmq/socket.hpp
@@ -607,7 +607,7 @@ public:
      */
     boost::system::error_code shutdown(shutdown_type what,
                                        boost::system::error_code & ec) {
-        return get_service().shutdown(get_implementation(), what, ec);
+        return get_service().descriptor_shutdown(get_implementation(), what, ec);
     }
 
     /** \brief Initiate shutdown of socket

